support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go

```diff
@@ -0,0 +1,204 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"path"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	svidv1 "github.com/spiffe/spire-api-sdk/proto/spire/api/server/svid/v1"
+	spiretypes "github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+	"k8s.io/client-go/util/retry"
+)
+
+const (
+	certDirMode   = os.FileMode(0o600)
+	certsFileMode = os.FileMode(0o644)
+	keyFileMode   = os.FileMode(0o600)
+	certsFileName = "tls.crt"
+	keyFileName   = "tls.key"
+)
+
+type SVIDConfig struct {
+	Cluster            string
+	Log                logrus.FieldLogger
+	Namespace          string
+	S                  svidv1.SVIDClient
+	TrustDomain        string
+	WebhookCertDir     string
+	WebhookServiceName string
+}
+
+type SVID struct {
+	c            SVIDConfig
+	certHalfLife time.Time
+	id           spiffeid.ID
+	notAfter     time.Time
+}
+
+// NewSVID creates a new SVID struct and creates the SPIFFE ID that will be used
+func NewSVID(ctx context.Context, config SVIDConfig) (*SVID, error) {
+	err := os.MkdirAll(config.WebhookCertDir, certDirMode)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := spiffeid.New(config.TrustDomain, "k8s-workload-registrar", config.Cluster, "webhook")
+	if err != nil {
+		return nil, fmt.Errorf("unable to generate SPIFFE ID: %w", err)
+	}
+
+	return &SVID{
+		c:  config,
+		id: id,
+	}, nil
+}
+
+// MintSVID requests the SPIRE Server to mint a new SVID for the webhook
+func (e *SVID) MintSVID(ctx context.Context, key crypto.Signer) (err error) {
```

Comment:
This is ok for now, but there is some shared state on the SVID type that is could end up being accessed by multiple goroutines based solely on the exported functions from the SVID type. I don't necessarily think we need to fix it now, but thought I'd put it on your radar in case this code is changed in future refactor.

Reply:
Yes good point. We should also make sure only one instance of the SVIDRotator is running.